[HOTFIX S3C 9.1.0.1] CLDSRV-173 CLDSRV-170 CLDSRV-177 S3C-5390

jonathan-gramain opened the pull request:
Backporting fixes that were dropped accidentally in regular maintenance branches

watsaqat committed:
S3C-5390 s3api head-object with part-number 1 on empty file fails: httpCode 416

(cherry picked from commit 4f3195a6ca6990636bd268b45aa509ae82e137bc)
(cherry picked from commit 8f3e7376647445e0ee2c9493c7991263f810ad8d)
(cherry picked from commit 64334db65a7aa862ee36808702cd3b59f25b3bf5)

jonathan-gramain committed:
bugfix: CLDSRV-170 skip orphan cleanup in UploadPart[Copy]

Do not delete orphan data in UploadPart/UploadPartCopy on overwrite
iff a CompleteMPU of the target MPU is already in progress.

This is to prevent a race condition where a CompleteMPU is running
while UploadPart is uploading a part for the same MPU.

It leaves an orphan in storage since only one of the upload data will
be present in the finished MPU, but the window is limited to the
CompleteMPU execution and should only occur when there are retries of
UploadPart due to prior stuck requests, or with broken clients
misusing the MPU API, so it should be acceptable.

Implementation details:

- set a flag in the MPU overview key when starting the CompleteMPU
  process, before listing the parts from metadata to construct the
  manifest

- in UploadPart/UploadPartCopy, after the part metadata is written and
  if the same part already existed, re-fetch the MPU overview key to
  check the flag: if set, skip the deletion of the old data of this
  part, since the CompleteMPU process in progress may choose either
  part data depending on the exact timing of the listing vs. the
  part overwrite.

(cherry picked from commit 84961115183cec724c7a9946dea4889f24c734a7)
(cherry picked from commit c2dbbfa0087078b793c0a92cab62dfe68436e7d1)
(cherry picked from commit d4063e157a1715bb3220fc7feab0749140b384b4)

watsaqat committed:
remove .only

(cherry picked from commit b4725aa03241f296808bdf53f747ec1b1306eb3f)
(cherry picked from commit af95fea311249dd1d81de891265229a246276aa4)
(cherry picked from commit e6bda3460b86e13d0b2e29c8b32436ca550d6e7a)

nicolas2bert committed:
CLDSRV-173 DeleteMarkers created by Lifecycle should not be replicated

(cherry picked from commit d30430a81c2e0bbba48a014e907f95272c50d303)
(cherry picked from commit a48138453815ffb86f9ad31c31e3ec8818ed28d7)

jonathan-gramain committed:
bugfix: CLDSRV-177 fix crash with empty object replication

Fix a case of crash when a replication occurs with an empty object
over a non-empty object.

It is not clear how this happens in practice but there can be some
corner cases with race conditions between object replication and
versioning getting suspended on the target bucket at the same time,
that could lead to this situation, as the check between replication
configuration and actual replication is not atomic.

(cherry picked from commit a4e8cbebe6e7c1be1ac7efdbe1ce6c666a0ad5a2)
(cherry picked from commit 108d1c920f946cb7f4237a5fcb0093345483da4d)
(cherry picked from commit 2b32ec61633bf3f30402fc7ea38152352af4511b)

jonathan-gramain committed:
[7.10] CLDSRV-177 add missing test helper checkObjectData

(cherry picked from commit 413ebe743ce3522efe433362a7b19212c1083729)
(cherry picked from commit a1c4420eab4e30e4388c0212bd2d8f82ae28d204)
(cherry picked from commit 6f32ebb2ce5f73d914c037a27279b516b9f4bb58)

jonathan-gramain committed:
CLDSRV-173 CLDSRV-170 CLDSRV-177 S3C-5390 bump hotfix version

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jonathan-gramain committed:
CLDSRV-177 fix backbeat func tests for CLDSRV-290

Since CLDSRV-290, backbeat routes functional tests need to pass a
versionId in the query string in the `/_/backbeat/metadata` route,
hence the original tests from CLDSRV-177 needed to be updated.

(cherry picked from commit fa562ae85a9e4200c33faa9e8830c7717d057888)

jonathan-gramain commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `hotfix/7.70.11`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.4`
* `development/8.5`
* `development/8.6`
* `development/8.7`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `hotfix/7.70.11`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.4`
* `development/8.5`
* `development/8.6`
* `development/8.7`



Please check the status of the associated issue CLDSRV-173.

Goodbye jonathan-gramain.
